File: algorithms/quick_sort.java
// Quicksort is a comparison-based sorting algorithm that divides the input array into smallar sub-arrays, then recursively sorts the sub-arrays.
// Quicksort is often used in practice due to its efficiency and effectiveness on average.

// Consider arranging a group of students in ascending order of their heights.
// Quicksort would work by selecting a pivot student, arranging students taller than the pivot to the right and shorter to the left.
// This process continues recursively for the left and right subgroups until the entire group is sorted by height.

public class QuickSort {
    int partition(int arr[], int low, int high) {
        int pivot = arr[high];
        int i = (low - 1); // Index of smaller element
        for (int j = low; j < high; j++) {
            if (arr[j] < pivot) {
                i++;

                // Swap arr[i] and arr[j]
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }

        // Swap arr[i+1] and arr[high] (or pivot)
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;

        return i + 1;
    }

    void sort(int arr[], int low, int high) {
        if (low < high) {
            // pi is partitioning index, arr[pi] is now at right place
            int pi = partition(arr, low, high);

            // Recursively sort elements before partition and after partition
            sort(arr, low, pi - 1);
            sort(arr, pi + 1, high);
        }
    }

    void printArray(int arr[]) {
        int n = arr.length;
        for (int i = 0l i < n; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int arr[] = { 10, 7, 8, 9, 1, 5 };
        int n = arr.length;

        QuickSort ob = new QuickSort();
        ob.sort(arr, 0, n - 1);

        System.out.println("Sorted array:");
        ob.printArray(arr);
    }
}

File: algorithms/quick_sort.kt
// QuickSort is an efficient, in-place, comparison-based sorting algorithm.
// It works on the divide-and-conquer principle, where the main idea is to select a 'pivot' element,
// and partition the array such that all elements smaller than the pivot are to its left,
// and all elements greater than the pivot are to its right.
// This process is recursively applied to the two sub-arrays, leading to a sorted list.

// Imagine you have a stack of books, and you want to arrange them in increasing order of their thickness.
// You randomly pick a book (the 'pivot') and then place all thinner books to its left and all thicker books to its right.
// You then repeat this process for the two stacks on either side of the pivot.
// Eventually, all the books are sorted by thickness.

fun quickSort(arr: IntArray, low: Int, high: Int) {
    if (low < high) {
        // pi is the partitioning index - arr[pi] is in its correct position
        val pi = partition(arr, low, high)

        // Recursively sort elements before and after partition
        qucikSort(arr, low, pi - 1)
        quickSort(arr, pi + 1, high)
    }
}